### PR TITLE
[Circle CI] Add Danger and publish to npm

### PR DESCRIPTION
**Pull request by hramos**
# Test Plan

Run on Circle

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -84,52 +85,57 @@ defaults: &defaults
...
-          command: |
-            if [ -n "$CIRCLE_PR_NUMBER" ]; then
-              npm install github@0.2.4
-              cat <(echo eslint; npm run lint --silent -- --format=json; echo flow; npm run flow --silent -- check --json) | GITHUB_TOKEN="af6ef0d15709bc91d""06a6217a5a826a226fb57b7" CI_USER=$CIRCLE_PROJECT_USERNAME CI_REPO=$CIRCLE_PROJECT_REPONAME PULL_REQUEST_NUMBER=$CIRCLE_PR_NUMBER node bots/code-analysis-bot.js
-            else 
-              echo "Skipping code analysis."
-            fi
 
-  test-node-6:
+  test-javascript-node-6:
```
Comment (grabbou) [inline .circleci/config.yml]: `js` is more concise?

Comment (hramos) [reply .circleci/config.yml]: Sure. Feel free to edit the file on the PR (if you have access)

File: .circleci/config.yml [hunk 2]
```diff
@@ -84,52 +85,57 @@ defaults: &defaults
...
     docker:
       - image: circleci/node:8
     steps:
       - checkout
-      - run: npm install --no-package-lock
+      - run: 
+          name: Install Dependencies
+          command: |
+            npm config set spin=false
+            npm config set progress=false
```
Comment (grabbou) [inline .circleci/config.yml]: Can we [make a common step](https://github.com/hramos/react-native/blob/4c1ef4ed4dc8e0b168d53321c67778271a6bfe36/.circleci/config.yml#L57-L61) for `npm install` just like we did for cache restoration and use it like [this](https://github.com/hramos/react-native/blob/4c1ef4ed4dc8e0b168d53321c67778271a6bfe36/.circleci/config.yml#L301)? We seem to be repeating that over and over in many places.

File: .circleci/config.yml [hunk 3]
```diff
@@ -84,52 +85,57 @@ defaults: &defaults
...
+            npm install
       - run: |
           npm test -- --maxWorkers=2
           npm run lint
           npm run flow -- check
 
-  test-node-4:
+  test-javascript-node-4:
```
Comment (grabbou) [inline .circleci/config.yml]: I think `js` is more concise?

File: .circleci/config.yml [hunk 4]
```diff
@@ -346,15 +365,90 @@ jobs:
...
+      - run:
+          name: Analyze Code
+          command: |
+            if [ -n "$CIRCLE_PULL_REQUEST" ]; then
+              cat <(echo eslint; npm run lint --silent -- --format=json; echo flow; npm run flow --silent -- check --json) | GITHUB_TOKEN="af6ef0d15709bc91d""06a6217a5a826a226fb57b7" CI_USER=$CIRCLE_PROJECT_USERNAME CI_REPO=$CIRCLE_PROJECT_REPONAME PULL_REQUEST_NUMBER=$CIRCLE_PR_NUMBER node bots/code-analysis-bot.js
+            else
+              echo "Skipping code analysis."            
+            fi
+
+  publish-npm:
```
Comment (grabbou) [inline .circleci/config.yml]: Discussion here: https://github.com/facebook/react-native/pull/16348#discussion_r144476728

**Comment by grabbou** [comment]
I will also test this PR on my (feature branch). Historically, some builds were failing. Will wait with this once we finish the config itself.

File: .circleci/config.yml (hunk 1 above)
Comment (grabbou) [reply .circleci/config.yml]: Ah, true, will play around :)

File: .circleci/config.yml (hunk 2 above)
Comment (hramos) [reply .circleci/config.yml]: Agreed. I like what you did in the other PR.

**Pull request closed without merging**